### PR TITLE
[util] Standardize auto-generated code warning msg

### PR DESCRIPTION
**Pull request by sriyerg**
This PR attempts to provide a standalone Python method that returns a
warning message box indicating that the source is autogenerated.

The `get_autogen_banner()` takes two arguments - `command` and
`comment`. It uses `tabulate` to create a text box containing the
warning banner and the supplied command that wraps to ~70 characters and
is center-justified for a clean look. The `comment` which is the correct
style supported by the file is applied to all lines and the final text
is returned as a string. Below are some examples:
Example 1:
```python
# +---------------------------------------------------------------------+
# | ----- W A R N I N G: A U T O - G E N E R A T E D   C O D E !! ----- |
# | DO NOT EDIT THIS FILE DIRECTLY. IT HAS BEEN AUTO-GENERATED WITH THE |
# |                         FOLLOWING COMMAND:                          |
# |    util/topgen.py -t hw/top_earlgrey/data/top_earlgrey.hjson -o     |
# |                           hw/top_earlgrey                           |
# +---------------------------------------------------------------------+
```
Example 2:
```systemverilog
// +---------------------------------------------------------------------+
// | ----- W A R N I N G: A U T O - G E N E R A T E D   C O D E !! ----- |
// | DO NOT EDIT THIS FILE DIRECTLY. IT HAS BEEN AUTO-GENERATED WITH THE |
// |                         FOLLOWING COMMAND:                          |
// |   ./util/dvsim/dvsim.py hw/top_earlgrey/dv/chip_sim_cfg.hjson -i    |
// |         chip_plic_all_irqs --waves -v h --local --run-only          |
// +---------------------------------------------------------------------+
```

It will be great to standardize on how we indicate the source has been
autogenerated, considering we have a lot of scripts that autogenerate
lots of things in our codebase. Happy to hear thoughts and opinions on 
improving how this should look further.

Signed-off-by: Srikrishna Iyer <sriyer@google.com>

### CODE REVIEW

**Comment by rswarbrick** [comment]
Looks like a good idea. I'm not so sure about the `main` function, though: I'd have thought we'd want to dissuade people from running the script and pasting the results into their templates :-)

**Comment by sriyerg** [comment]
> Looks like a good idea. I'm not so sure about the `main` function, though: I'd have thought we'd want to dissuade people from running the script and pasting the results into their templates :-)

Agreed - the `main()` is temporary, just until we finalize how should this look.

**Comment by sriyerg** [comment]
> Also, I feel like the box/text centering is unnecessary; among other things, it makes it quite hard to copy-paste the command, which ideally should just be on one long, un-wrapped line.
Very well, sounds good. Updated.
> 
> Finally, I'm not sure this is an exhaustive list of affected scripts, and we probably want this to be discoverable to stop it from backsliding... @timothytrippel you will have opinions. =)
Agreed -will make this discoverable and apply to all scripts once we are happy with how this looks.

**Comment by msfschaffner** [comment]
Thanks for doing this, @sriyerg. I think this makes sense.

Regarding long commands:
I agree that putting them onto one long line makes copy pasting more convenient. However, this will probably create issues with our various SV linters that check for excessive line lengths. Now, it can be argued whether it makes sense to run the linters on generated code (there are pros and cons), or whether the 100 char limit makes sense. 

But given that we do not have to copy paste these commands on a regular basis, but rather in rare debug cases, I would be ok with wrapping the commands automatically (maybe we can automatically add backslashes `\`?)...

**Comment by rswarbrick** [comment]
I'm with @msfschaffner and would keep the line breaks: if we have to run the command manually, it's not a big deal to remove a couple of `#` characters :-) (And, hopefully, we've got CI checks that ensure that we'll never need to run it manually anyway!)

File: hw/ip/alert_handler/data/alert_handler.hjson.tpl [hunk 1]
```diff
@@ -2,14 +2,16 @@
...
-#
-# Parameter (given by Python tool)
-#  - n_alerts:    Number of alert sources
-#  - esc_cnt_dw:  Width of escalation counter
-#  - accu_cnt_dw: Width of accumulator
-#  - async_on:    Enables asynchronous sygnalling between specific alert RX/TX pairs
-#  - n_classes:   Number of supported classes (leave this at 4 at the moment)
+${autogen_banner}
+
+// ALERT_HANDLER register template
```
Comment (mcy) [inline hw/ip/alert_handler/data/alert_handler.hjson.tpl:7]: Why are these // comments now?

File: hw/ip/flash_ctrl/data/flash_ctrl.hjson [hunk 2]
```diff
@@ -1,8 +1,7 @@
 // Copyright lowRISC contributors.
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
-
-
+//
```
Comment (mcy) [inline hw/ip/flash_ctrl/data/flash_ctrl.hjson:4]: There seems to be a lot of random noise in this PR that seems unrelated to the change?

**Review by mcy — COMMENTED** [review]
I think the commit diffs are a bit noisy and make it hard to tell what's changed. Can you blow away the second commit and rebuiild it to only contain changes due to the new tool?

To Michael's comment, it may be worth trying to crop down the size of the printed commands, but I'm not thrilled. I would also consider avoiding making all of the lines be as long as they are (with the dashes and whatnot) since they make wrapping sad in other places, too.

I'd take a look at something lighter weight, such as what manticore does: https://github.com/lowRISC/manticore/blob/master/fuzz/gen/manticore_protocol_capabilities_DeviceCapabilities__req_from_wire.rs#L5.

Finally, regarding `main()`: I don't think it's worth keeping while we're discussing the format. If the format is such a huge deal, maybe we should break it off as a separate discussion... but that seems like a poor use of time.

File: hw/ip/alert_handler/data/alert_handler.hjson.tpl (hunk 1 above)
Comment (sriyerg) [reply hw/ip/alert_handler/data/alert_handler.hjson.tpl:7]: Fixed an inconsistency here (Hjson supports both styles, but `//` is used more predominantly).

File: hw/ip/flash_ctrl/data/flash_ctrl.hjson (hunk 2 above)
Comment (sriyerg) [reply hw/ip/flash_ctrl/data/flash_ctrl.hjson:4]: It is related. Flash ctrl stuff is autogenerated with another script which has not been updated.  This should get cleaned up in subsequent PRs.

**Comment by sriyerg** [comment]
> I think the commit diffs are a bit noisy and make it hard to tell what's changed. Can you blow away the second commit and rebuiild it to only contain changes due to the new tool?
> 
Ok. 

> To Michael's comment, it may be worth trying to crop down the size of the printed commands, but I'm not thrilled. I would also consider avoiding making all of the lines be as long as they are (with the dashes and whatnot) since they make wrapping sad in other places, too.
> 
I agree with others - having to manually run the commands will be very rare, so splitting it on multiple lines does not seem like a huge deal. 

> I'd take a look at something lighter weight, such as what manticore does: https://github.com/lowRISC/manticore/blob/master/fuzz/gen/manticore_protocol_capabilities_DeviceCapabilities__req_from_wire.rs#L5.
> 
> Finally, regarding `main()`: I don't think it's worth keeping while we're discussing the format. If the format is such a huge deal, maybe we should break it off as a separate discussion... but that seems like a poor use of time.

Yes, I did say im my previous comment that I will remove it.

**Comment by sriyerg** [comment]
In the latest force-push, I addressed most of the comments. The warning banner now looks like this:
```
# ---------------------------------------------------------------------
# ------ W A R N I N G: A U T O - G E N E R A T E D   C O D E !! ------
# DO NOT EDIT THIS FILE DIRECTLY. GENERATED WITH THE FOLLOWING COMMAND:
# .util/uvmdvgen/uvmdvgen.py dma -e -eo hw/ip/dma/dv
# ---------------------------------------------------------------------


// ---------------------------------------------------------------------
// ------ W A R N I N G: A U T O - G E N E R A T E D   C O D E !! ------
// DO NOT EDIT THIS FILE DIRECTLY. GENERATED WITH THE FOLLOWING COMMAND:
// util/topgen.py -t hw/top_earlgrey/data/top_earlgrey.hjson -o
// hw/top_earlgrey
// ---------------------------------------------------------------------
```

The input command is wrapped so that the linters are happy.

**Comment by timothytrippel** [comment]
> I think the commit diffs are a bit noisy and make it hard to tell what's changed. Can you blow away the second commit and rebuiild it to only contain changes due to the new tool?
> 

For future reference, sometimes with these sorts of things I find it easier to split into ~2+ PRs (similar to how I did with the IRQ DIFs):
    -1st PR: just the tooling and proposed format
    -2nd+ PR: updating templates and re-generating code with new templates.

> To Michael's comment, it may be worth trying to crop down the size of the printed commands, but I'm not thrilled. I would also consider avoiding making all of the lines be as long as they are (with the dashes and whatnot) since they make wrapping sad in other places, too.
> 
> I'd take a look at something lighter weight, such as what manticore does: https://github.com/lowRISC/manticore/blob/master/fuzz/gen/manticore_protocol_capabilities_DeviceCapabilities__req_from_wire.rs#L5.
> 
> Finally, regarding `main()`: I don't think it's worth keeping while we're discussing the format. If the format is such a huge deal, maybe we should break it off as a separate discussion... but that seems like a poor use of time.

**Review by timothytrippel — APPROVED** [review]
Looks like you split up the PRs too, thanks!

LGTM.

**Comment by mcy** [comment]
> Our tools are quite customizable - topgen for example takes different arguments to generate code in specific ways. This does not seem feasible.

I think I was unclear. I'm suggesting that all of the `-----` characters and spaces between letters are unnecessary; keep it simple. This is going to become irrelevant once we bring Bazel online, so let's not do anything fancy.

File: util/autogen_banner.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+AUTOGEN_BANNER = [
+    "-------------------------------------------------------------------------",
+    "-------- W A R N I N G: A U T O - G E N E R A T E D   C O D E !! --------",
+    "DO NOT EDIT THIS FILE DIRECTLY. GENERATED WITH THE FOLLOWING COMMAND:",
+    "-------------------------------------------------------------------------",
+]
+
+
+def get_autogen_banner(command: str, comment=None) -> str:
```
Comment (mcy) [inline util/autogen_banner.py:21]: Why not extract `command` from `sys`?

File: util/autogen_banner.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    command is a fully formatted string representing what command was used
+    to auto-generate the source.
+    comment is the style of comment supported by the file type.
+    """
+    command_lines = wrap(command.strip(), width=len(AUTOGEN_BANNER[0]))
+    banner = AUTOGEN_BANNER[:-1] + command_lines + [AUTOGEN_BANNER[-1]]
+    return apply_comment(banner, comment)
+
+
+def get_license_banner(comment=None) -> str:
```
Comment (mcy) [inline util/autogen_banner.py:33]: Why does this script care about the license header?

File: util/autogen_banner.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+
+def get_autogen_banner(command: str, comment=None) -> str:
+    """Returns a commented out auto-generated code warning banner.
+
+    command is a fully formatted string representing what command was used
+    to auto-generate the source.
+    comment is the style of comment supported by the file type.
+    """
+    command_lines = wrap(command.strip(), width=len(AUTOGEN_BANNER[0]))
```
Comment (mcy) [inline util/autogen_banner.py:28]: You need to configure this a bit more carefully, since textwrap is not actually intended for code. You want:
```python
command_lines = wrap(
  command,
  width=MAX_LEN,
  drop_whitespace=True,
  break_long_words=False,
  break_on_hyphens=False,
  initial_indent='  '
)
```
You then want to include backslashes like Michael requested:
```python
for i in range(0, len(command_lines)-1):
  command_lines += ' \\'
```

You will want `MAX_LEN` to be its own constant.

File: util/autogen_banner.py (hunk 3 above)
Comment (mcy) [reply util/autogen_banner.py:21]: Why is `comment` untyped? If you're going to type part of the file, type all of it.

File: util/autogen_banner.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    banner = AUTOGEN_BANNER[:-1] + command_lines + [AUTOGEN_BANNER[-1]]
```
Comment (mcy) [inline util/autogen_banner.py:29]: It probably makes sense to, instead of making assumptions about the size of the banner, to include a token in the banner so that you can `.format()` it. E.g. `AUTOGEN_HEADER.format(command=command)`.

(I would also suggest making the constant a string; apply comment would better apply to a string than to a list.)

File: util/autogen_banner.py (hunk 4 above)
Comment (sriyerg) [reply util/autogen_banner.py:33]: Some files are auto-generated right from within the Python script. Those will need to add license banner as well. See:
https://github.com/lowRISC/opentitan/blob/366021d6f1bacab43ac310d80f2984487857db93/util/topgen.py#L44

File: util/autogen_banner.py (hunk 5 above)
Comment (sriyerg) [reply util/autogen_banner.py:28]: Adding backslash breaks the CI. GCC does not like it:
```
FAILED: sw/device/tests/autogen/971261f@@plic_all_irqs_test_lib@sta/plic_all_irqs_test.c.o 
/opt/buildcache/riscv/bin/riscv32-unknown-elf-gcc -Isw/device/tests/autogen/971261f@@plic_all_irqs_test_lib@sta  -I../../s/sw/device/tests/autogen -I/home/vsts/work/1/s -I/home/vsts/work/1/a/build-out -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Werror -std=c11 -g -nostdinc -isystem/home/vsts/work/1/s/sw/device/lib/base/freestanding -fno-asynchronous-unwind-tables -fno-common -ffunction-sections -Wimplicit-fallthrough -Wswitch-default -Wno-error=unused-function -Wno-unused-parameter -Wno-sign-compare -Wno-missing-field-initializers -Os -fvisibility=hidden -Wa,--no-pad-sections -Wstrict-prototypes -march=rv32imc -mabi=ilp32 -mcmodel=medany -MD -MQ 'sw/device/tests/autogen/971261f@@plic_all_irqs_test_lib@sta/plic_all_irqs_test.c.o' -MF 'sw/device/tests/autogen/971261f@@plic_all_irqs_test_lib@sta/plic_all_irqs_test.c.o.d' -o 'sw/device/tests/autogen/971261f@@plic_all_irqs_test_lib@sta/plic_all_irqs_test.c.o' -c ../../s/sw/device/tests/autogen/plic_all_irqs_test.c
../../s/sw/device/tests/autogen/plic_all_irqs_test.c:8:1: error: multi-line comment [-Werror=comment]
    8 | // util/topgen.py -t hw/top_earlgrey/data/top_earlgrey.hjson \
      | ^
cc1: all warnings being treated as errors
```

File: util/autogen_banner.py (hunk 3 above)
Comment (sriyerg) [reply util/autogen_banner.py:21]: > Why not extract `command` from `sys`?

I will let @tjaychen & @eunchan answer this - in `topgen` the command is hardcoded in various places seemingly because some of the switches that take default values wont be present in the incantation, but it is useful to tell the reader how to change those default values if they wish to.

File: util/autogen_banner.py (hunk 6 above)
Comment (sriyerg) [reply util/autogen_banner.py:29]: The only assumption made here is that the command (now wrapped into a list of lines) needs to be inserted before the last line. Given that `AUTOGEN_BANNER` is file local constant, it should be ok to make such assumptions. 

The comment needs to be applied to all lines of the text. It is hence, better to keep the constant as a list rather than string. Otherwise we need to split it on newlines and join again.

File: util/autogen_banner.py (hunk 5 above)
Comment (rswarbrick) [reply util/autogen_banner.py:28]: It should work if you add another trailing space (because then the C code is essentially "escaping" a space, rather than the newline). It's a bit ugly, though :-/

**Comment by sriyerg** [comment]
> > Our tools are quite customizable - topgen for example takes different arguments to generate code in specific ways. This does not seem feasible.
> 
> I think I was unclear. I'm suggesting that all of the `-----` characters and spaces between letters are unnecessary; keep it simple. This is going to become irrelevant once we bring Bazel online, so let's not do anything fancy.

IMO having an outline is useful. I will make the other text a bit simpler. I will let others weigh in too. 
Not sure how is bazel going to render this obsolete (maybe for SW) - keeping the generated HW sources checked in was a long long discussion that we've already had (and have agreed to it).

File: util/autogen_banner.py (hunk 5 above)
Comment (sriyerg) [reply util/autogen_banner.py:28]: It doesn't seem worth the trouble - lets just skip it?

**Comment by sriyerg** [comment]
> > > Our tools are quite customizable - topgen for example takes different arguments to generate code in specific ways. This does not seem feasible.
> > 
> > 
> > I think I was unclear. I'm suggesting that all of the `-----` characters and spaces between letters are unnecessary; keep it simple. This is going to become irrelevant once we bring Bazel online, so let's not do anything fancy.
> 

For something that you think is going to be irrelevant, we sure are having a lot of discussions :)

File: util/autogen_banner.py (hunk 5 above)
Comment (msfschaffner) [reply util/autogen_banner.py:28]: yeah let's just skip it then.

File: util/autogen_banner.py (hunk 3 above)
Comment (moidx) [reply util/autogen_banner.py:21]: Update as `comment: str`. It seems that the consuming code can use an explicit  value.

File: util/autogen_banner.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    """
+    return apply_comment(LICENSE_BANNER, comment)
+
+
+def apply_comment(text: list, comment=None) -> str:
+    """Applies comment to a text paragraph.
+
+    The returned string terminates in a newline.
+    """
+    if comment:
```
Comment (moidx) [inline util/autogen_banner.py:46]: with `comment: str` you probably can get rid of the conditional. If needed you can strip spaces in comment, but it seems unnecessary.

File: util/autogen_banner.py (hunk 5 above)
Comment (moidx) [reply util/autogen_banner.py:28]: backslashes have been problematic in autogen comments. I don't see consistent results from auto formatting. Lets leave that out unless there is a good way to handle them. The config options proposed by @mcy seem reasonable to me.

File: util/autogen_banner.py (hunk 6 above)
Comment (moidx) [reply util/autogen_banner.py:29]: nit: Most autogen messages I have seen lately are as simple as this. The prefix, suffix lines are not as common anymore. This code would look simpler without the line suffix. Leaving as nit.

```cc
// THIS FILE HAS BEEN GENERATED, DO NOT EDIT MANUALLY.
// Command: {command}
```

**Review by moidx — APPROVED** [review]

**Review by msfschaffner — APPROVED** [review]
LGTM from my side

File: util/autogen_banner.py (hunk 6 above)
Comment (sriyerg) [reply util/autogen_banner.py]: Done.

File: util/autogen_banner.py (hunk 5 above)
Comment (sriyerg) [reply util/autogen_banner.py]: Done.

File: util/autogen_banner.py (hunk 3 above)
Comment (sriyerg) [reply util/autogen_banner.py]: Done.

**Comment by sriyerg** [comment]
Addressed all comments in the latest push.

**Review by mcy — APPROVED** [review]
